L69-cpp-core-event-manager.md

```diff
@@ -0,0 +1,282 @@
+  // ConsistentCancel. Callers of ConsistentCancel need to be cautious to avoid 
+  // deadlock.
+  virtual void ScheduleAlarmIn(gpr_timespec deadline, std::function<void(bool)> c, Priority priority, AlarmKey* key = nullptr) = 0;
+
+  // Adds a closure to the EventManager, which will be run at the earliest possible 
+  // time. The TaskKey pointer is an output parameter which can be used to cancel 
+  // the closure at a later time (currently not supported in 
+  // BaseEventManagerInterface). Note that the closure should not be blocking or 
+  // long-running, as it would block the worker thread in the EventManager from 
+  // polling the file descriptors.
+  virtual void RunFunction(std::function<void()> handler, Priority priority, TaskKey* key = nullptr) = 0;
+
+  struct Task {
+    std::function<void()> handler;
+    Priority priority;
+    TaskKey* key;
+    int internal_added_queue;
+  };
+  // Adds multiple closures to the event management system all at once.
+  virtual void RunFunctions(Task* tasks, int length) = 0;
+
+  // Cancels an alarm which has been added to the EventManager via 
+  // ScheduleAlarmIn(). This function will run the closure with false in the 
+  // calling thread if it successfully cancels the alarm.
+  virtual void ConsistentCancel(const AlarmKey& key) = 0;
+
+  // Cancels an alarm as ConsistentCancel(), but wait if the task is currently 
+  // running.
+  virtual void BlockingConsistentCancel(const AlarmKey& key) = 0;
+
+  // These two methods adjust a refcount of the EventManager, preventing it from 
+  // shutdown if the refcount is positive. ShutdownRef() should be called before 
+  // adding the first closure to the EventManager, and ShutdownUnref() can be 
+  // called after removing the last closure from the EventManager. Also, if you 
+  // need to keep the EventManager serving before receiving an RPC response, you 
+  // need to call ShutdownRef() before sending the RPC, and invoke ShutdownUnref() 
+  // when the response arrives.
+  virtual void ShutdownRef() = 0;
+  virtual void ShutdownUnref() = 0;
+};
+```
+
+The BaseEventManagerInterface is platform agnostic in order to accommodate the different polling models on different platforms (e.g. POSIX vs Windows).
+
+A specific interface EpollEventManagerInterface is derived from BaseEventManagerInterface for all the epoll-based polling models.
+
+```cpp
+// Represents a file descriptor on epoll-capable platforms.
+class EpollDescriptorInterface {
+ public:
+  virtual ~EpollDescriptorInterface();
+
+  // Returns the underlying file descriptor.
+  virtual int fd() const = 0;
+
+  // Returns the EventManager object where this Descriptor is registered.
+  virtual EpollEventManagerInterface* event_manager() const = 0;
+
+  // Executes all the registered notification closures, and closes the underlying 
+  // file descriptor.
+  virtual void Close() = 0;
+
+  // Returns whether the underlying file descriptor is closed or not.
+  virtual bool IsClosed() const = 0;
```

Comment:
const member function?

Reply:
Not sure if I missed anything. It's already const, is it not?

Reply:
Yeah. Not sure what I meant myself.